contracts/GenArt721CoreV3_Engine_Flex.sol

```diff
@@ -0,0 +1,2278 @@
+        // record contracts starting project ID
+        // casting-up is safe
+        startingProjectId = uint256(_startingProjectId);
+        _updateProviderSalesAddresses(
+            _renderProviderAddress,
+            _renderProviderAddress,
+            _platformProviderAddress,
+            _platformProviderAddress
+        );
+        _updateRandomizerAddress(_randomizerContract);
+        // set AdminACL management contract as owner
+        _transferOwnership(_adminACLContract);
+        // initialize default base URI
+        _updateDefaultBaseURI(
+            string.concat(
+                "https://token.artblocks.io/",
+                toHexString(address(this)),
+                "/"
+            )
+        );
+        // initialize next project ID
+        _nextProjectId = _startingProjectId;
+        emit PlatformUpdated(FIELD_NEXT_PROJECT_ID);
+        // register contract as an Engine contract
+        IEngineRegistryV0(_engineRegistryContract).registerContract(
+            address(this),
+            CORE_VERSION,
+            CORE_TYPE
+        );
+    }
+
+    /**
+     * @notice Updates preferredIPFSGateway to `_gateway`.
+     */
+    function updateIPFSGateway(string calldata _gateway) public {
+        _onlyAdminACL(this.updateIPFSGateway.selector);
+        preferredIPFSGateway = _gateway;
+        emit GatewayUpdated(ExternalAssetDependencyType.IPFS, _gateway);
+    }
+
+    /**
+     * @notice Updates preferredArweaveGateway to `_gateway`.
+     */
+    function updateArweaveGateway(string calldata _gateway) public {
+        _onlyAdminACL(this.updateArweaveGateway.selector);
+        preferredArweaveGateway = _gateway;
+        emit GatewayUpdated(ExternalAssetDependencyType.ARWEAVE, _gateway);
+    }
+
+    /**
+     * @notice Locks external asset dependencies for project `_projectId`.
+     */
+    function lockProjectExternalAssetDependencies(uint256 _projectId) external {
+        _onlyUnlockedProjectExternalAssetDependencies(_projectId);
+        _onlyArtistOrAdminACL(
+            _projectId,
+            this.lockProjectExternalAssetDependencies.selector
+        );
+        projects[_projectId].externalAssetDependenciesLocked = true;
+        emit ProjectExternalAssetDependenciesLocked(_projectId);
+    }
+
+    /**
+     * @notice Returns external asset dependency for project `_projectId` at index `_index`.
+     * If the dependencyType is ONCHAIN, the `data` field will contain the extrated bytecode data and `cid`
+     * will be an empty string. Conversly, for any other dependencyType, the `data` field will be an empty string
+     * and the `bytecodeAddress` will point to the zero address.
+     */
+    function projectExternalAssetDependencyByIndex(
+        uint256 _projectId,
+        uint256 _index
+    ) public view returns (ExternalAssetDependencyWithData memory) {
+        ExternalAssetDependencyType _dependencyType = projects[_projectId]
```

Comment:
nit: can we eliminate some redundancy (and some redundant hashing and follow pattern like:
```
// first load pointer to external asset dependency in storage
ExternalAssetDependency storage dependency = projects[_projectId].externalAssetDependencies[_index];
// then lookup properties
address bytecodeAddress = dependency.bytecodeAddress
// ... etc.
```